### PR TITLE
Rewrite SILCombiner::propagateConcreteTypeOfInitExistential.

### PR DESCRIPTION
**Pull request by atrick**
Fixes <rdar://40555427> [SR-7773]:
SILCombiner::propagateConcreteTypeOfInitExistential fails to full propagate type
substitutions.

This rewrite fixes several fundamental bugs in the SILCombiner optimization that
propagates concrete types. In particular, the pass needs to handle:

- Arguments of callee Self type in non-self position.
- Indirect and direct return values of Self type.
- Types that indirectly depend on Self within callee function signature.
- Protocol composition existentials.
- All of the above need to work for protocol extensions as well as witness methods.
- For protocol extensions, conformance lookup should be based on the existential's conformance list.

Additionally, the optimization should not depend on a SILFunction's DeclContext,
which is not serialized. (In fact, we should prevent SIL passes from using
DeclContext). Furthermore, the code needs to be expressed in a way that one can
reason about correctness and invariants.

The root cause of these bugs is that SIL passes are written based on untested
assumptions of Swift type system. A SIL pass needs to handle all verifiable SIL
input because passes need to be composable. Bail-out logic can be added to
simplify the design; however, _the bail-out logic itself cannot make any
assumptions about the language or type system_ that aren't clearly and
explicitly enforced in the SIL verifier. This is a common mistake and major
source of bugs.

I created as many unit tests as I reasonably could to prevent this code from
regressing. Creating enough unit tests to cover all corner cases that were
broken in the original code would be intractable. But the code has been
simplified such that many corner cases disappear.

This opens up some oportunity for generalizing the optimization and eliminating
special cases. However, I want this PR to be limited to fixing correctness
issues only. In the long term, it would be preferable to replace this
optimization entirely with a much more powerful general type propagation pass.

### CODE REVIEW

**Comment by atrick** [comment]
@slavapestov Normally I would not introduce an optimization rewrite late in a release branch. However, the rewritten code is lower risk than leaving any of the existing code in place. Removing the optimization altogether would introduce the risk of sudden regressions in a release that we don't have time to fix.

The best way to review this I think is to checkout the branch and read through these functions (there's nothing in the original code worth looking at other than what was copied into the ConcreteExistentialInfo constructor):

  SILCombiner::createApplyWithConcreteType
  SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite AI, WitnessMethodInst *WMI)
  SILCombiner::propagateConcreteTypeOfInitExistential(FullApplySite AI)

**Comment by atrick** [comment]
@rajbarik this might trivially conflict with your PRs. I left `findInitExistential` in the code for your convenience, but needed to slightly change the signature. If ConcreteExistentialInfo can be used in all cases instead, then we could remove it.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
@swift-ci benchmark.

**Comment by atrick** [comment]
The -Onone benchmark regressions and improvements are all spurious. None of this code is touched at -Onone.

**Comment by rajbarik** [comment]
Thanks. I will take a look after it is merged.

**Comment by atrick** [comment]
@rajbarik or @slavapestov It would be good to get a review on this, not just to unblock @rajbarik, but also to get bug fixes on 4.2 sooner. If you don't have time let me know so I can recruit someone for the review.

**Comment by atrick** [comment]
@swift-ci test.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (rajbarik) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: @atrick Does this handle non-self arguments? One part of my optimizations was to add this functionality https://github.com/apple/swift/pull/13991.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (rajbarik) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: At some point, we should merge what is there in PR https://github.com/apple/swift/pull/13991 to handle both self and non-self concrete type propagation. The concrete type can be obtained using a number of techniques including protocolconformance analysis, global_addr pattern match and init_existential pattern match. Makes sense?

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (atrick) [reply lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: I'm limiting this PR to fixing correctness and trying hard to generalize the code without generalizing the current behavior. 

The code in this function subsumes what used to be in multiple `propagateConcreteTypeOfInitExistential` functions (for both witness and protocol extension functions).

Relative to your PR #13991, it is one level above your simple `createApplyWithConcreteType` and could probably be reused by your `propagateConcreteTypeOfInitExistentialToAllApplyArgs` by generalizing the behavior. I think these PRs are mostly complementary and merging them later will be an additional code quality improvement.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (atrick) [reply lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: Yes, it should be fully generalized. I just need to separate out the bug fix which can hopefully be merged onto a release branch.

**Comment by atrick** [comment]
@slavapestov do you have time to review this for 4.2?

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (rajbarik) [reply lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: +1

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (rajbarik) [reply lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: Sounds good.

**Review by slavapestov — APPROVED** [review]
Thanks!

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test source compatibility.

**Comment by atrick** [comment]
SCK fails on master:
FAIL_CoreStore_3.2_BuildXcodeWorkspaceScheme_CoreStoreDemo_generic-platform-iOS.log
Assertion failed: ((HadError || !M.is<SourceFile*>() || M.get<SourceFile*>()->ASTStage < SourceFile::TypeChecked) && "UnresolvedDot" "in wrong phase"), function walkToExprPre, file swift/include/swift/AST/ExprNodes.def, line 101.

File: lib/SILOptimizer/Utils/Existential.cpp
Comment (rajbarik) [inline lib/SILOptimizer/Utils/Existential.cpp]: @atrick @slavapestov Could you decouple the body from the constructor and invoke it separately via an API? "findInitExistential" is not the only way to determine concrete types. For example, I am using ProtocolConformance Analysis to determine concrete types. I am planning on using the same ConcreteExistentialInfo struct and reuse the createApplyWithConcreteType. Let me know if you are busy and it sounds good to you, I can try to decouple them.

Comment (atrick) [reply lib/SILOptimizer/Utils/Existential.cpp]: @rajbarik All suggestions are welcome. I'm not dogmatic about how ConcreteExistentialInfo is constructed or accessed. I like to start with minimal design and develop outward. If you have a rough prototype, and want quick feedback, just point me to a branch or gist.

First thing to consider: do all the members of ConcreteExistentialInfo make sense for all the SIL patterns you're handling. It's fine to have a few optional fields:
- ConcreteTypeDef is null to when ConcreteType is not an opened archetype
- You could make InitExistential null to indicate the existential wasn't found.

Anything much beyond that and it may be better to have a discrimiated union or a smaller more general Info composing a larger one.

At any rate, I think it's fine if you optionally pass in a PCA to the ConcreteExistentialInfo construction, or even define a second constructor. If that's awkward for some reason you could move to a free standing or static function to create the Info.

Comment (rajbarik) [reply lib/SILOptimizer/Utils/Existential.cpp]: @atrick Thanks. I am currently adding a new constructor for CEI with values computed from PCA and this constructor will take a subset of the values currently present in the CEI. We can later think about a more tighter integration. However, one problem I am running into is regarding the following code in method createApplyWithConcreteType:
```
if (CEI.isCopied && !canReplaceCopiedSelf(Apply, CEI.InitExistential, DA))
```
When we use PCA, there is no InitExistential and no isCopied. So, the question is, can this check be moved to isValid() function in CEI or prepended with a check on InitExistential not being null?  Do you have any recommendation?

Comment (rajbarik) [reply lib/SILOptimizer/Utils/Existential.cpp]: Please keep in mind that a check on InitExistential being not null, is minimal in terms of code changes :)

Comment (atrick) [reply lib/SILOptimizer/Utils/Existential.cpp]: The safe thing to do is to bail out under the `isCopied` case if we didn't find the init_existential.
`if (isCopied && (!InitExistential || !canReplaceCopiedSelf(Apply, InitExistential)`
 I'm sure we could do better but if you don't care about that case and never hit it then why deal with it?

**Pull request merged**
